File: MiniDOMBase/DMD/Sistema/EMailSys/CEmailAccount.vb
Imports System.Net.Mail 'importo il Namespace
Imports System.Net.Sockets
Imports minidom
Imports minidom.Databases
Imports minidom.Anagrafica

 

Partial Public Class Sistema

    <Serializable> _
    Public Class CEmailAccount
        Inherits DBObject

        Private m_Attivo As Boolean
        Private m_AccountType As String
        Private m_AccountName As String
        Private m_POPServer As String
        Private m_POPPort As Integer
        Private m_POPUserName As String
        Private m_POPPassword As String
        Private m_POPUseSSL As Boolean
        Private m_LastSync As Date?

        Public Sub New()
            Me.m_Attivo = True
            Me.m_AccountName = ""
            Me.m_AccountType = "POP3"
            Me.m_POPServer = "localhost"
            Me.m_POPPort = 110
            Me.m_POPUserName = ""
            Me.m_POPPassword = ""
            Me.m_POPUseSSL = False
            Me.m_LastSync = Nothing
        End Sub

        Public Sub New(ByVal userName As String, ByVal password As String)
            Me.New()
            Me.m_POPUserName = Trim(userName)
            Me.m_POPPassword = password
            Me.m_AccountName = Me.m_POPUserName
        End Sub

        ''' <summary>
        ''' Restituisce o imposta la data e l'ora dell'ultima sincronizzazione avvenuta con il server
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property LastSync As Date?
            Get
                Return Me.m_LastSync
            End Get
            Set(value As Date?)
                Dim oldValue As Date? = Me.m_LastSync
                If (DateUtils.Compare(value, oldValue) = 0) Then Exit Property
                Me.m_LastSync = value
                Me.DoChanged("LastSync", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta un valore booleano che indica se l'account è attivo
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Attivo As Boolean
            Get
                Return Me.m_Attivo
            End Get
            Set(value As Boolean)
                If (Me.m_Attivo = value) Then Exit Property
                Me.m_Attivo = value
                Me.DoChanged("Attivo", value, Not value)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il tipo dell'account (POP3 e IMAP)
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property AccountType As String
            Get
                Return Me.m_AccountType
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_AccountType
                If (oldValue = value) Then Exit Property
                Me.m_AccountType = value
                Me.DoChanged("AccountType", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il nome del server POP
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property POPServer As String
            Get
                Return Me.m_POPServer
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_POPServer
                If (oldValue = value) Then Exit Property
                Me.m_POPServer = value
                Me.DoChanged("POPServer", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta la porta del server POP
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property POPPort As Integer
            Get
                Return Me.m_POPPort
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_POPPort
                If (oldValue = value) Then Exit Property
                Me.m_POPPort = value
                Me.DoChanged("POPPort", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta la porta del server POP
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property POPUseSSL As Boolean
            Get
                Return Me.m_POPUseSSL
            End Get
            Set(value As Boolean)
                If (Me.m_POPUseSSL = value) Then Exit Property
                Me.m_POPUseSSL = value
                Me.DoChanged("POPUseSSL", value, Not value)
            End Set
        End Property


        ''' <summary>
        ''' Restituisce o imposta il nome utente per l'accesso al server POP
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property POPUserName As String
            Get
                Return Me.m_POPUserName
            End Get
            Set(value As String)
                value = Strings.Trim(value)
                Dim oldValue As String = Me.m_POPUserName
                If (oldValue = value) Then Exit Property
                Me.m_POPUserName = value
                Me.DoChanged("POPUserName", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta la password per l'accesso al server POP
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property POPPassword As String
            Get
                Return Me.m_POPPassword
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_POPPassword
                If (oldValue = value) Then Exit Property
                Me.m_POPPassword = value
                Me.DoChanged("POPPassword", value, oldValue)
            End Set
        End Property

        ''' <summary>
        ''' Restituisce o imposta il nome dell'account
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property AccountName As String
            Get
                Return Me.m_AccountName
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_AccountName
                If (oldValue = value) Then Exit Property
                Me.m_AccountName = value
                Me.DoChanged("AccountName", value, oldValue)
            End Set
        End Property
             

        Public Overrides Function GetModule() As CModule
            Return Nothing
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_EmailAccounts"
        End Function

        Protected Overrides Function LoadFromRecordset(ByVal reader As DBReader) As Boolean
            Me.m_Attivo = reader.Read("Attivo", Me.m_Attivo)
            Me.m_AccountName = reader.Read("AccountName", Me.m_AccountName)
            Me.m_AccountType = reader.Read("AccountType", Me.m_AccountType)
            Me.m_POPServer = reader.Read("POPServer", Me.m_POPServer)
            Me.m_POPPort = reader.Read("POPPort", Me.m_POPPort)
            Me.m_POPUserName = reader.Read("POPUserName", Me.m_POPUserName)
            Me.m_POPPassword = reader.Read("POPPassword", Me.m_POPPassword)
            Me.m_POPUseSSL = reader.Read("POPUseSSL", Me.m_POPUseSSL)
            Me.m_LastSync = reader.Read("LastSync", Me.m_LastSync)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(ByVal writer As DBWriter) As Boolean
            writer.Write("Attivo", Me.m_Attivo)
            writer.Write("AccountName", Me.m_AccountName)
            writer.Write("AccountType", Me.m_AccountType)
            writer.Write("POPServer", Me.m_POPServer)
            writer.Write("POPPort", Me.m_POPPort)
            writer.Write("POPUserName", Me.m_POPUserName)
            writer.Write("POPPassword", Me.m_POPPassword)
            writer.Write("POPUseSSL", Me.m_POPUseSSL)
            writer.Write("LastSync", Me.m_LastSync)

            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "Attivo" : Me.m_Attivo = XML.Utils.Serializer.DeserializeBoolean(fieldValue)
                Case "AccountName" : Me.m_AccountName = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "AccountType" : Me.m_AccountType = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "POPServer" : Me.m_POPServer = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "POPPort" : Me.m_POPPort = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "POPUserName" : Me.m_POPUserName = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "POPPassword" : Me.m_POPPassword = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "POPUseSSL" : Me.m_POPUseSSL = XML.Utils.Serializer.DeserializeBoolean(fieldValue)
                Case "LastSync" : Me.m_LastSync = XML.Utils.Serializer.DeserializeDate(fieldValue)

                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("Attivo", Me.m_Attivo)
            writer.WriteAttribute("AccountName", Me.m_AccountName)
            writer.WriteAttribute("AccountType", Me.m_AccountType)
            writer.WriteAttribute("POPServer", Me.m_POPServer)
            writer.WriteAttribute("POPPort", Me.m_POPPort)
            writer.WriteAttribute("POPUserName", Me.m_POPUserName)
            writer.WriteAttribute("POPPassword", Me.m_POPPassword)
            writer.WriteAttribute("POPUseSSL", Me.m_POPUseSSL)
            writer.WriteAttribute("LastSync", Me.m_LastSync)

            MyBase.XMLSerialize(writer)
        End Sub

        Protected Friend Overrides Function GetConnection() As Databases.CDBConnection
            Return Databases.APPConn
        End Function

        Sub Load()
            Dim tbl As minidom.Databases.CDBTable = APPConn.Tables.GetItemByKey(Me.GetTableName)
            Dim col As minidom.Databases.CDBEntityField = tbl.Fields.GetItemByKey("LastSync")
            If (col Is Nothing) Then
                col = tbl.Fields.Add("LastSync", GetType(Date))
                col.Create()
            End If

            Dim reader As DBReader
            Dim dbSQL As String = "SELECT * FROM [" & Me.GetTableName & "] ORDER BY [ID] ASC"
            reader = New DBReader(APPConn.Tables(Me.GetTableName), dbSQL)
            If reader.Read Then
                APPConn.Load(Me, reader)
            End If
            reader.Dispose()
        End Sub

        Protected Overrides Sub DoChanged(propName As String, Optional newVal As Object = Nothing, Optional oldVal As Object = Nothing)
            MyBase.DoChanged(propName, newVal, oldVal)
            EMailer.doConfigChanged(New System.EventArgs)
        End Sub
 


    End Class


End Class

File: MiniDOMBase/DMD/Sistema/EMailSys/Accounts/CEmailAccount.cs
using System;
using DMD;
using DMD.Databases;
using DMD.XML;
using minidom;
using static minidom.Sistema;
using minidom.repositories;

namespace minidom
{
    public partial class Sistema
    {

        /// <summary>
        /// Account configurato su un server web
        /// </summary>
        [Serializable]
        public class CEmailAccount
            : Databases.DBObject
        {
            private bool m_Attivo;
            private string m_AccountType;
            private string m_AccountName;
            private string m_POPServer;
            private int m_POPPort;
            private string m_POPUserName;
            private string m_POPPassword;
            private bool m_POPUseSSL;
            private DateTime? m_LastSync;

            /// <summary>
            /// Costruttore
            /// </summary>
            public CEmailAccount()
            {
                m_Attivo = true;
                m_AccountName = "";
                m_AccountType = "POP3";
                m_POPServer = "localhost";
                m_POPPort = 110;
                m_POPUserName = "";
                m_POPPassword = "";
                m_POPUseSSL = false;
                m_LastSync = default;
            }

            /// <summary>
            /// Costruttore
            /// </summary>
            /// <param name="userName"></param>
            /// <param name="password"></param>
            public CEmailAccount(string userName, string password) : this()
            {
                m_POPUserName = DMD.Strings.Trim(userName);
                m_POPPassword = password;
                m_AccountName = m_POPUserName;
            }

            /// <summary>
            /// Restituisce o imposta la data e l'ora dell'ultima sincronizzazione avvenuta con il server
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public DateTime? LastSync
            {
                get
                {
                    return m_LastSync;
                }

                set
                {
                    var oldValue = m_LastSync;
                    if (DMD.DateUtils.Compare(value, oldValue) == 0)
                        return;
                    m_LastSync = value;
                    DoChanged("LastSync", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta un valore booleano che indica se l'account è attivo
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool Attivo
            {
                get
                {
                    return m_Attivo;
                }

                set
                {
                    if (m_Attivo == value)
                        return;
                    m_Attivo = value;
                    DoChanged("Attivo", value, !value);
                }
            }

            /// <summary>
            /// Restituisce o imposta il tipo dell'account (POP3 e IMAP)
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string AccountType
            {
                get
                {
                    return m_AccountType;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_AccountType;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_AccountType = value;
                    DoChanged("AccountType", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta il nome del server POP
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string POPServer
            {
                get
                {
                    return m_POPServer;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_POPServer;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_POPServer = value;
                    DoChanged("POPServer", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la porta del server POP
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public int POPPort
            {
                get
                {
                    return m_POPPort;
                }

                set
                {
                    int oldValue = m_POPPort;
                    if (oldValue == value)
                        return;
                    m_POPPort = value;
                    DoChanged("POPPort", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la porta del server POP
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public bool POPUseSSL
            {
                get
                {
                    return m_POPUseSSL;
                }

                set
                {
                    if (m_POPUseSSL == value)
                        return;
                    m_POPUseSSL = value;
                    DoChanged("POPUseSSL", value, !value);
                }
            }


            /// <summary>
            /// Restituisce o imposta il nome utente per l'accesso al server POP
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string POPUserName
            {
                get
                {
                    return m_POPUserName;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_POPUserName;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_POPUserName = value;
                    DoChanged("POPUserName", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la password per l'accesso al server POP
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string POPPassword
            {
                get
                {
                    return m_POPPassword;
                }

                set
                {
                    string oldValue = m_POPPassword;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_POPPassword = value;
                    DoChanged("POPPassword", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta il nome dell'account
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string AccountName
            {
                get
                {
                    return m_AccountName;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_AccountName;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_AccountName = value;
                    DoChanged("AccountName", value, oldValue);
                }
            }

            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Sistema.EMailer.MailAccounts;
            }

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_EmailAccounts";
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                m_Attivo = reader.Read("Attivo", m_Attivo);
                m_AccountName = reader.Read("AccountName", m_AccountName);
                m_AccountType = reader.Read("AccountType", m_AccountType);
                m_POPServer = reader.Read("POPServer", m_POPServer);
                m_POPPort = reader.Read("POPPort", m_POPPort);
                m_POPUserName = reader.Read("POPUserName", m_POPUserName);
                m_POPPassword = reader.Read("POPPassword", m_POPPassword);
                m_POPUseSSL = reader.Read("POPUseSSL", m_POPUseSSL);
                m_LastSync = reader.Read("LastSync", m_LastSync);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva sul db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("Attivo", m_Attivo);
                writer.Write("AccountName", m_AccountName);
                writer.Write("AccountType", m_AccountType);
                writer.Write("POPServer", m_POPServer);
                writer.Write("POPPort", m_POPPort);
                writer.Write("POPUserName", m_POPUserName);
                writer.Write("POPPassword", m_POPPassword);
                writer.Write("POPUseSSL", m_POPUseSSL);
                writer.Write("LastSync", m_LastSync);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara il db
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("Attivo", typeof(bool), 1);
                c = table.Fields.Ensure("AccountName", typeof(string), 255);
                c = table.Fields.Ensure("AccountType", typeof(string), 255);
                c = table.Fields.Ensure("POPServer", typeof(string), 255);
                c = table.Fields.Ensure("POPPort", typeof(int), 1);
                c = table.Fields.Ensure("POPUserName", typeof(string), 255);
                c = table.Fields.Ensure("POPPassword", typeof(string), 255);
                c = table.Fields.Ensure("POPUseSSL", typeof(bool), 1);
                c = table.Fields.Ensure("LastSync", typeof(DateTime), 1);
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.GetItemByName("idxStato");
                if (c is object)
                    c.Drop();
                c = table.Constraints.Ensure("idxAccountName", new string[] { "AccountName", "Stato" }, DBFieldConstraintFlags.PrimaryKey);
                c = table.Constraints.Ensure("idxParams", new string[] { "AccountType", "Attivo" }, DBFieldConstraintFlags.PrimaryKey);
                c = table.Constraints.Ensure("idxPOP", new string[] { "POPServer", "POPPort", "POPUseSSL", "LastSync"  }, DBFieldConstraintFlags.PrimaryKey);
                c = table.Constraints.Ensure("idxUser", new string[] { "POPUserName", "POPPassword" }, DBFieldConstraintFlags.PrimaryKey);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Attivo":
                        {
                            m_Attivo = DMD.XML.Utils.Serializer.DeserializeBoolean(fieldValue) == true;
                            break;
                        }

                    case "AccountName":
                        {
                            m_AccountName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "AccountType":
                        {
                            m_AccountType = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "POPServer":
                        {
                            m_POPServer = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "POPPort":
                        {
                            m_POPPort = (int)DMD.XML.Utils.Serializer.DeserializeInteger(fieldValue);
                            break;
                        }

                    case "POPUserName":
                        {
                            m_POPUserName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "POPPassword":
                        {
                            m_POPPassword = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "POPUseSSL":
                        {
                            m_POPUseSSL = DMD.XML.Utils.Serializer.DeserializeBoolean(fieldValue) == true;
                            break;
                        }

                    case "LastSync":
                        {
                            m_LastSync = DMD.XML.Utils.Serializer.DeserializeDate(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Seralizzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Attivo", m_Attivo);
                writer.WriteAttribute("AccountName", m_AccountName);
                writer.WriteAttribute("AccountType", m_AccountType);
                writer.WriteAttribute("POPServer", m_POPServer);
                writer.WriteAttribute("POPPort", m_POPPort);
                writer.WriteAttribute("POPUserName", m_POPUserName);
                writer.WriteAttribute("POPPassword", m_POPPassword);
                writer.WriteAttribute("POPUseSSL", m_POPUseSSL);
                writer.WriteAttribute("LastSync", m_LastSync);
                base.XMLSerialize(writer);
            }

           
        }
    }
}